src/backend/cpu/memory.cpp

```diff
@@ -56,29 +52,30 @@ void printMemInfo(const char *msg, const int device) {
 template<typename T>
 unique_ptr<T[], function<void(T *)>> memAlloc(const size_t &elements) {
     T *ptr = nullptr;
-
-    common::MemoryEventPair me = memoryManager().alloc(elements * sizeof(T), false);
-    if(me.e) me.e.enqueueWait(getQueue());
-    ptr = (T *)me.ptr;
+    buffer_info pair =
+        buffer_info(memoryManager().alloc(elements * sizeof(T), false));
+    event ev = event(pair.getEvent());
+    Event &e = getEvent(ev.get());
```

Comment:
I have similar concern about all `ev` objects created in all backends' memory.cpp files. Just leaving this comment instead of duplicating comments in files from all backends.

Reply:
@9prady9 — `af::event` here is a convenience to automatically free the `af_event` after it falls out of scope. We could remove it, but then the above would look like:
```c++
    buffer_info pair =
        buffer_info(memoryManager().alloc(elements * sizeof(T), false));
    af_event eventHandle = pair.getEvent(); // new
    Event &e = getEvent(eventHandle);
    if (e) e.enqueueWait(getQueue());
    ptr = (T *)pair.getPtr();
    af_release_event(eventHandle); // now we need this
    return unique_ptr<T[], function<void(T *)>>(ptr, memFree<T>);
```
releasing the event handle incurs a bit of overhead for the user and requires calling the C API directly.

> Event &e = getEvent(event(pair.getEvent()).get());

this won't work because using the `af::event` as a temporary leads to its immediate destruction when `getEvent` returns, which renders the underlying `detail::Event` invalid

> Perhaps `getEvent` can be renamed to `getEventHandle`

this `detail::Event` isn't a handle - `af_event` is the public handle - `detail::Event` is the actual underlying event which contains some `EventType e_` (e.g. a `cuEvent_t`), so I'm not sure we should rename. Right now there's: `af::event` (public C++ API), `af_event` (public C API handle), and `detail::Event` (internal implementation). Do you think removing `using Event = detail::Event` and referring to `Event` as `detail::Event` internally would this clearer?

Reply:
Most recent change addressed my concern now that event returned `getEventFromBufferInfoHandle` is being moved.